### PR TITLE
How to Develop ActiveRecord SQL Server Adapter with Pre-Installed MS SQL

### PR DESCRIPTION
**Pull request by lcreid**
This PR adds documentation to show how to use a pre-prepared Vagrant virtual machine for testing the ActiveRecord SQL Server Adapter. It follows an approach used by the Rails core team for Rails development -- meaning development of Rails itself, not development of applications built on Rails. Similarly, this is not for developing Rails applications on SQL Server, it's for working on the SQL Server Adapter itself.

Developers who don't have easy access the SQL Server may find it easier to contribute to the SQL Server Adapter if they follow the instructions in this PR.

I have used this approach for some debugging of Rails 6 support, and it seems to be working. 

I happy for any and all feedback. Let me know if you have any questions or comments.

### CODE REVIEW

**Comment by ttilberg** [comment]
This is great, I think it's clear that right now we need to make it easier for people to participate in the dev of this adapter. I hope this gets merged sooner than later in effort to overcome the Rails 6 challenge that has been simmering for quite some time.

Yesterday I took my first dive into dev against the adapter, and I was pretty successful mirroring the strategy used for CI with a few small tweaks -- namely:
- In the Dockerfile:
  - Copying the gem related files and doing bundle install before copying the app to cache it
  - Installing `less` so that Pry works as expected
- A few other small workflow things

I found the workflow very workable and surprisingly straightforward on an Ubuntu box. The image used for CI already has the test dbs set up, and of course is on parity with the main CI testing pipeline. The only problems were related to the image being built in a way that isn't friendly to rebuilding. I know several people who might jam on Docker, but not Vagrant.

If I were to add notes about the Docker workflow, would you see it best as adding to your PR, or a separate PR against this project? I think it would make sense to have this effort coordinated.


```diff
ttilberg@timdesktop:~/dev/dev/activerecord-sqlserver-adapter$ git diff Dockerfile
diff --git a/Dockerfile b/Dockerfile
index 8b18d5c..882db15 100644
--- a/Dockerfile
+++ b/Dockerfile
@@ -1,14 +1,23 @@
 ARG TARGET_VERSION=2.6.3

 FROM wpolicarpo/activerecord-sqlserver-adapter:${TARGET_VERSION}
+# Add less to support Pry:
+RUN apk add less

 ENV WORKDIR /activerecord-sqlserver-adapter

 RUN mkdir -p $WORKDIR
 WORKDIR $WORKDIR

-COPY . $WORKDIR
+# Copy the bare necessities for bundling
+COPY VERSION Gemfile* *.gemspec ./
+COPY lib/active_record/connection_adapters/sqlserver/version.rb ./lib/active_record/connection_adapters/sqlserver/

+# Cache the bundle
 RUN bundle install --jobs `expr $(cat /proc/cpuinfo | grep -c "cpu cores") - 1` --retry 3

+# Bring the rest of the code over
+COPY . $WORKDIR
+
 CMD ["sh"]
```

**Comment by lcreid** [comment]
My guess is that one page with two sections, one for Docker and one for Vagrant, would be the best. I do Vagrant because that's what Rails core does, but I know a lot of people have more experience with Docker, so ideally we would cover both. And, as you say, Docker fits with our CI.

I don't have any experience with this type of situation. I don't know what would be the easiest way for us to include your Docker notes in the same document. I suppose you could fork mine and submit a PR back here, or a PR to my repo. Whatever you think would work best, go for it.

Of course, if you're a committer on this project just merge my PR and work from there. :-)

Whatever approach you take, I support it. I think we should cover both Docker and Vagrant.

**Comment by wpolicarpo** [comment]
@lcreid thanks for this. I'd rather have this on a wiki page if you don't mind?

The wiki area is currently public, which means anyone can make updates. We'll change that, but in the meantime can I ask you to move this to the wiki area and link the README to there?

**Comment by wpolicarpo** [comment]
@lcreid forget what I said before. Let's create a `docs` folder and move `DEVELOPING_WITH_VAGRANT.md` to there. Can you do that?

**Comment by ttilberg** [comment]
@wpolicarpo That sounds good, we should also add a note in the readme to find that file: I often miss those. 

I'll add a separate PR for Docker notes that we can refine and address separately.

**Comment by lcreid** [comment]
Thanks, @wpolicarpo . I can change the PR to create a `docs` folder and move `DEVELOPING_WITH_VAGRANT.md`.

**Comment by lcreid** [comment]
@wpolicarpo updated this PR as requested.

File: docs/DEVELOPING_WITH_VAGRANT.md [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+
+You will need about 10 GB free space for each instance of the virtual machine that you want to use. You also need 4 GB of RAM in addition to whatever else you're running on your host.
+
+## How To Build The Virtual Machine
+
+Building the virtual machine is this easy:
+
+    host $ mkdir mssql # Or whatever name suits you
+    host $ cd mssql
+    host $ vagrant jadesystems/rails-jade-18-04-mssql
```
[the remark this answers is not in the document]
Comment (aidanharan) [reply docs/DEVELOPING_WITH_VAGRANT.md]: When I run try to create the virtual machine I get a "File source must be specified." error. Any idea why? I'm running Vagrant 2.2.7 and VirtuaBox 6.1.4. Maybe I shouldn't be running `vagrant init ...`?

```
$ vagrant init jadesystems/rails-jade-18-04-mssql
A `Vagrantfile` has been placed in this directory. You are now
ready to `vagrant up` your first virtual environment! Please read
the comments in the Vagrantfile as well as documentation on
`vagrantup.com` for more information on using Vagrant.

$ vagrant up
Bringing machine 'default' up with 'virtualbox' provider...
There are errors in the configuration of this machine. Please fix
the following errors and try again:

File provisioner:
* File source must be specified.
```

[the remark this answers is not in the document]
Comment (lcreid) [reply docs/DEVELOPING_WITH_VAGRANT.md]: `vagrant init` is correct. What operating system are you running on your laptop/desktop?

[the remark this answers is not in the document]
Comment (aidanharan) [reply docs/DEVELOPING_WITH_VAGRANT.md]: I'm running macOS Mojave (v10.14.6).

[the remark this answers is not in the document]
Comment (aidanharan) [reply docs/DEVELOPING_WITH_VAGRANT.md]: Upgraded and I'm having the same issue with macOS Catalina (v10.15.4).

**Comment by aidanharan** [comment]
_(This is a repeat of a thread that doesnt automatically display because the original issue was resolved. So starting new thread about it here.)_

When I run try to create the virtual machine I get a "File source must be specified." error. Any idea why? 

```
$ vagrant init jadesystems/rails-jade-18-04-mssql
A `Vagrantfile` has been placed in this directory. You are now
ready to `vagrant up` your first virtual environment! Please read
the comments in the Vagrantfile as well as documentation on
`vagrantup.com` for more information on using Vagrant.

$ vagrant up
Bringing machine 'default' up with 'virtualbox' provider...
There are errors in the configuration of this machine. Please fix
the following errors and try again:

File provisioner:
* File source must be specified.
```

Issue is happening on:
- macOS Mojave (v10.14.6)
- - Vagrant 2.2.7 & VirtuaBox 6.0.16
- - Vagrant 2.2.7 & VirtuaBox 6.1.4
- macOS Catalina (v10.15.4)
- - Vagrant 2.2.7 & VirtuaBox 6.1.4

**Comment by lcreid** [comment]
@aidanharan thanks for the details on this problem. I believe I've fixed the problem of the Vagrant box on macOS but I need to test it. I'll push a new version of the Vagrant box as soon as I can. Thanks for your patience.

**Comment by lcreid** [comment]
There is a new version of the Vagrant box, 6.2.1. I hope it fixes the issue. Unfortunately, I wasn't able to get my hands on a Mac today, so I couldn't test it. If it works for you, please let me know. @aidanharan

**Comment by aidanharan** [comment]
The latest version of VirtualBox is v6.1.4 (https://www.virtualbox.org/wiki/Downloads) and latest Vagrant is v2.2.7 (https://www.vagrantup.com/downloads.html). I'm not sure what you mean by "6.2.1". 

@lcreid What version of OS, VirtualBox and Vagrant is working for you when you use "jadesystems/rails-jade-18-04-mssql"?

**Comment by lcreid** [comment]
6.2.1 is the version of the `rails-jade-18-04-mssql` box itself. That's where the problem is. 

You have to update your Vagrant box. Note that this will destroy your existing VM, so you'll lose the contents of your MS-SQL database. You don't lose anything in the shared directories where your source code is.

To update the box, you'll need another ~1.5 GB space on your drive. Change to the directory where the box is and run:

```
vagrant box update
vagrant box destroy
vagrant box up
```

FWIW, here's the Vagrant documentation: https://www.vagrantup.com/docs/boxes/versioning.html.

**Comment by aidanharan** [comment]
Great that worked. @lcreid Could you change the PR so that it merges into the `6-0-dev` branch? I've only tested it for Rails 6 and thats what you used it for too afaik.

In the future I might look at updating the `Vagrant` file so that it uses the standard Ubuntu image and the required software is installed using scripts.

**Comment by aidanharan** [comment]
@lcreid Like the rails-dev-box project there is now one for developing the SQL Server adapter. I forked it off the rails-dev-box and then added changes from this PR. https://github.com/rails-sqlserver/activerecord-sqlserver-adapter-dev-box

This PR can probably be closed now.

**Comment by wpolicarpo** [comment]
I'd like to keep this MR but using the `dev-box` that was created. @lcreid would you be able to change the README pointing at https://github.com/rails-sqlserver/activerecord-sqlserver-adapter-dev-box instead of using a predefined image?

I guess `docs/DEVELOPING_WITH_VAGRANT.md` can be removed as the dev-box has a comprehensive README already, unless you have something to add?

**Comment by lcreid** [comment]
I hope this is what you wanted me to do, @wpolicarpo . Let me know if you need anything else.

**Comment by wpolicarpo** [comment]
Thank you so much @lcreid.

**Pull request merged**